src/backend/cdb/cdbtm.c

```diff
@@ -2370,16 +2372,51 @@ currentGxactWriterGangLost(void)
 	return MyTmGxactLocal->writerGangLost;
 }
 
+/*
+ * Add read only segments into MyTmGxactLocal->readOnlySegments.
+ */
+void
+addToReadOnlySegments(int segindex)
+{
+	MemoryContext oldContext;
+
+	if (!isCurrentDtxActivated())
+		return;
+
+	/* skip if all segdbs are in the list */
+	if (list_length(MyTmGxactLocal->readOnlySegments) >= getgpsegmentCount())
+		return;
+
+	/* entry db is just a reader, will not involve in two phase commit */
+	if (segindex == -1)
+		return;
+
+	oldContext = MemoryContextSwitchTo(TopTransactionContext);
+
+	/* skip if record already */
+	if (bms_is_member(segindex, MyTmGxactLocal->readOnlySegmentsMap) || 
+		bms_is_member(segindex, MyTmGxactLocal->dtxSegmentsMap))
+	{
+		MemoryContextSwitchTo(oldContext);
+		return;
+	}
```

Comment:
```suggestion
	/* skip if record already */
	if (bms_is_member(segindex, MyTmGxactLocal->readOnlySegmentsMap) || 
		bms_is_member(segindex, MyTmGxactLocal->dtxSegmentsMap))
	{
		return;
	}
	oldContext = MemoryContextSwitchTo(TopTransactionContext);
```

`bms_is_member()` doesn't call `palloc()`. We don't need to switch the memory context for it, right?

Reply:
ok